### ISSUE TITLE
Media Embed Automatic Embeds: Per provider configuration

### ISSUE THREAD

**Issue by tony** [ckeditor/ckeditor5#9816]
## 📝 Provide a description of the new feature

_What is the expected behavior of the proposed feature?_

### Per-provider configuration of auto embeds

Allow [Media provider](https://ckeditor.com/docs/ckeditor5/latest/features/media-embed.html#media-providers) configuration to specify whether they should be automatically embedded or not.

### Optional: Auto embed pattern filtering
Optionally allow specifying regex captures (e.g. youtube URL variations or a more permissive regex, capture a full iframe html block being pasted)

e.g. When adding or [overriding an existing provider](https://ckeditor.com/docs/ckeditor5/latest/features/media-embed.html#overriding-media-providers) like `youtube`, allow specifying a list of patterns, e.g.

```html
<iframe width="560" height="315" src="https://www.youtube.com/embed/IOvqN23Sr4o" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>```
```

### Locations
Package: [media-embed/mediaembed
](https://ckeditor.com/docs/ckeditor5/latest/api/module_media-embed_mediaembed.html) [`MediaEmbed`](https://ckeditor.com/docs/ckeditor5/latest/api/module_media-embed_mediaembed-MediaEmbed.html), [docs](https://ckeditor.com/docs/ckeditor5/latest/features/media-embed.html)
Module: [`AutoMediaEmbed`](https://ckeditor.com/docs/ckeditor5/latest/api/module_media-embed_automediaembed-AutoMediaEmbed.html)'s [Automatic embedding on paste](https://ckeditor.com/docs/ckeditor5/latest/features/media-embed.html#automatic-media-embed-on-paste)
File: [`automediaembed.js`](https://github.com/ckeditor/ckeditor5/blob/32639c726d7863eea0eb357db9bb3985c3d85d52/packages/ckeditor5-media-embed/src/automediaembed.js#L21-L26)

### Related

> We would likely also want to specify an ignore list of domains we never want to auto-embed. https://github.com/ckeditor/ckeditor5/issues/1300#issue-369663310

---

If you'd like to see this feature implemented, add a 👍 reaction to this post.

**Comment by Reinmar** [comment]
Hi! Apologies for the late reply.

I'm trying to understand what's the use case here. Why do you need this option?

**Comment by tony** [comment]
@Reinmar 

It's been a bit of time! I will try to jog my memory.

This wasn't a show stopper for us. We can close.

Backstory in case anyone bumps upon this in the future:

There was a specific desire we had where we wanted some domains to automatically paste (e.g. YouTube) but other providers we allow, they'd paste via clicking the button.  There wasn't a way to this without the PR, since there was no way to opt-out of automatic embedding. It was hard coded.

Do you think its okay to have all embedded regexes automatically paste without a way to control it?

**Comment by neongreen** [comment]
> where we wanted some domains to automatically paste (e.g. YouTube) but other providers we allow, they'd paste via clicking the button. There wasn't a way to this without the PR, since there was no way to opt-out of automatic embedding. It was hard coded.

Exactly this.

We are using embedly, and they can generate a decent-looking link preview for pretty much any link (think "link bookmarks" in Notion). They also have a constantly updating list of providers they support. Ideally we just want to allow any link in the "insert media" button — embedly will handle it.

However, we definitely don't want all links to turn into embed widgets on paste. We actually rolled this out and users are frustrated.

FWIW, having a Nimbus-style embed menu would solve this problem:

![image](https://user-images.githubusercontent.com/1523306/133660137-671de5a3-5e0a-4d76-80af-1561d42f9cd8.png)

But until then we would like to be able to configure autoembeds to fire only for specific providers (YouTube, etc) where we know that most of the time users want those to be autoembedded.

**Comment by CKEditorBot** [comment]
There has been no activity on this issue for the past two years. We've marked it as stale and will close it in 30 days. We understand it may be relevant, so if you're interested in the solution, leave a comment or reaction under this issue.

**Comment by Witoso** [comment]
Still relevant, rel: #1300.

**Comment by neongreen** [comment]
I ended up patching installed CKEditor in-place with https://www.npmjs.com/package/patch-package.

My patch is here:

```diff
diff --git a/node_modules/@ckeditor/ckeditor5-media-embed/src/automediaembed.js b/node_modules/@ckeditor/ckeditor5-media-embed/src/automediaembed.js
index e1863d8..ae91eb4 100644
--- a/node_modules/@ckeditor/ckeditor5-media-embed/src/automediaembed.js
+++ b/node_modules/@ckeditor/ckeditor5-media-embed/src/automediaembed.js
@@ -101,6 +101,12 @@ export default class AutoMediaEmbed extends Plugin {
             urlRange.detach();
             return;
         }
+        const media = mediaRegistry._getMedia(url);
+        // Do not embed if 'disableAutoEmbed'
+        if (media.disableAutoEmbed) {
+          urlRange.detach();
+          return;
+        }
         const mediaEmbedCommand = editor.commands.get('mediaEmbed');
         // Do not anything if media element cannot be inserted at the current position (#47).
         if (!mediaEmbedCommand.isEnabled) {
diff --git a/node_modules/@ckeditor/ckeditor5-media-embed/src/mediaembedconfig.d.ts b/node_modules/@ckeditor/ckeditor5-media-embed/src/mediaembedconfig.d.ts
index c911cda..3c964df 100644
--- a/node_modules/@ckeditor/ckeditor5-media-embed/src/mediaembedconfig.d.ts
+++ b/node_modules/@ckeditor/ckeditor5-media-embed/src/mediaembedconfig.d.ts
@@ -275,4 +275,8 @@ export interface MediaEmbedProvider {
      * is `true`, the rendering function **will always** be used for the media in the editor data output.
      */
     html?: (match: RegExpMatchArray) => string;
+    /**
+     * Patched-in. Disable auto embeds, while still allowing embeds via the "Media" button.
+     */
+    disableAutoEmbed?: boolean;
 }
diff --git a/node_modules/@ckeditor/ckeditor5-media-embed/src/mediaregistry.d.ts b/node_modules/@ckeditor/ckeditor5-media-embed/src/mediaregistry.d.ts
index 76a39dc..4e8ba0c 100644
--- a/node_modules/@ckeditor/ckeditor5-media-embed/src/mediaregistry.d.ts
+++ b/node_modules/@ckeditor/ckeditor5-media-embed/src/mediaregistry.d.ts
@@ -55,7 +55,7 @@ export default class MediaRegistry {
      * @param url The URL of the media.
      * @returns The `Media` instance or `null` when there is none.
      */
-    private _getMedia;
+    public _getMedia(url: string): Media;
     /**
      * Tries to match `url` to `pattern`.
      *
diff --git a/node_modules/@ckeditor/ckeditor5-media-embed/src/mediaregistry.js b/node_modules/@ckeditor/ckeditor5-media-embed/src/mediaregistry.js
index 5423070..d3dfd74 100644
--- a/node_modules/@ckeditor/ckeditor5-media-embed/src/mediaregistry.js
+++ b/node_modules/@ckeditor/ckeditor5-media-embed/src/mediaregistry.js
@@ -81,7 +81,9 @@ export default class MediaRegistry {
             for (const subPattern of pattern) {
                 const match = this._getUrlMatches(url, subPattern);
                 if (match) {
-                    return new Media(this.locale, url, match, previewRenderer);
+                    const media = new Media(this.locale, url, match, previewRenderer);
+                    if (definition.disableAutoEmbed) media.disableAutoEmbed = true;
+                    return media; 
                 }
             }
         }
@@ -119,7 +121,7 @@ export default class MediaRegistry {
  *
  * It can be rendered to the {@link module:engine/view/element~Element view element} and used in the editing or data pipeline.
  */
-class Media {
+export class Media {
     constructor(locale, url, match, previewRenderer) {
         this.url = this._getValidUrl(url);
         this._locale = locale;
```

It must be saved into `patches/@ckeditor+ckeditor5-media-embed+39.0.0.patch`.

**Comment by CKEditorBot** [comment]
There has been no activity on this issue for the past year. We've marked it as stale and will close it in 30 days. We understand it may still be relevant, so if you're interested in the solution, leave a comment or reaction under this issue.

**Comment by CKEditorBot** [comment]
There has been no activity on this issue for the past year. We've marked it as stale and will close it in 30 days. We understand it may still be relevant, so if you're interested in the solution, leave a comment or reaction under this issue.

**Comment by tony** [comment]
Still seems like an appropriate idea.
